dspy/functional/typed_predictor_signature.py

```diff
@@ -0,0 +1,130 @@
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:
+                field.validate_default = False
+
+            input_field = InputField(desc=field.description)
+            dspy_fields[field_name] = (field.annotation, input_field)
+
+        for field_name, field in pydantic_class_for_dspy_output_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:   
+                field.validate_default = False
+
+            if field.default is PydanticUndefined:
+                raise ValueError(
+                    f"Field '{field_name}' has no default value. Required fields must have a default value. "
+                    "Change the field to be Optional or specify a default value."
+                )
+            
+            output_field = OutputField(desc=field.description if field.description else "")
+            dspy_fields[field_name] = (field.annotation, output_field)
+
+            instructions += f"When extracting '{field_name}':\n"
+            instructions += f"If it is not mentioned in the input fields, return: '{field.default}'. "
+
+            examples = field.examples
+            if examples:
+                quoted_examples = [f"'{example}'" for example in examples]
+                instructions += f"Example values are: {', '.join(quoted_examples)} etc. "
+
+            if field.metadata:
+                constraints = [meta for meta in field.metadata if 'Validator' not in str(meta)]
+                if field.json_schema_extra and 'invalid_value' in field.json_schema_extra:
+                    instructions += f"If the extracted value does not conform to: {constraints}, return: '{field.json_schema_extra['invalid_value']}'."
```

Comment:
Detecting missing and invalid values is a "must-have" requirement for production-quality apps. Here, I am basically trying to avoid the multiple retries that the framework makes when the extracted value does not match the constraints specified in the Pydantic BaseModel. 

The LM may be incorrectly extracting the parameter or the user has specified an invalid value. In both cases, the system should not hallucinate the closest value that matches the actual input. Rather, it should flag the invalid field and let the application handle the error with a proper error-correction workflow ("Provided value was invalid. It must be... Here are some examples... Did you mean...?")